### DISCUSSION TITLE
Starting from tty. Hyprland runs fine, but I am not at systemd graphical-session.target

### DISCUSSION THREAD

**Discussion by sitedyno** [hyprwm/Hyprland#7766] [General]
This means desktop-portal-hyprland doesn't auto start. I'm on arch & up to date, how can I become graphical?

**Comment by MahouShoujoMivutilde** [comment]
Depends on how complex you want to get.

There is [hyprland-systemd.desktop](https://github.com/hyprwm/Hyprland/blob/main/example/hyprland-systemd.desktop ) with [hyprland-session.service](https://github.com/hyprwm/Hyprland/blob/main/example/hyprland-session.service) in examples.

I don't see why it wouldn't work, though if you e.g. change your PATH somewhere in your shell's `~/.profile` without updating systemd environment after (`systemctl --user import-environment PATH ...`; `dbus-update-activation-environment --systemd PATH ...`), hyprland (as it runs as a systemd service here), won't see a new PATH, and your bindings that call custom scripts e.g. in `~/.local/bin` won't work.

You don't have to run hyprland as a service though.

In the simplest case you want some `.target` (but as we can see above apparently `.service` works too) that has

```ini
# ~/.config/systemd/user/hyprland-session.target
[Unit]
# ...
BindsTo=graphical-session.target
# ...
```

to be launched when hyprland is ready (WAYLAND_DISPLAY etc is set, systemd / dbus environment is updated, so all your waybars won't crash on launch).

 (you can't start `graphical-session.target` directly)

e.g. 

```ini
exec-once = sleep 5s && systemctl --user start hyprland-session.target
```

(sleep is for variables to be updated)

And

```ini
exec-shutdown = systemctl --user stop hyprland-session.target
```

so that all units with `PartOf=graphical-session.target` will stop. See man [systemd.special](https://www.freedesktop.org/software/systemd/man/latest/systemd.special.html#graphical-session.target).

That's the gist of it.

In practice, you might not want this delay, and would prefer to update variables yourself (or with some other script).

For that there is [uwsm](https://github.com/Vladimir-csp/uwsm), which does, seemingly, everything, but is huge (haven't tried it). Or you could look at how sway-systemd's [session.sh](https://github.com/alebastr/sway-systemd/blob/main/src/session.sh) and [units](https://github.com/alebastr/sway-systemd/tree/main/units) do it, and adapt for hyprland, which is what I did. 

Basically

1. change names of units (duh)
2. split `session.sh` in two scripts: 
2.1. one that sets the variables and starts the session (cleaning up if failed / terminated) - to be launched with `exec-once`
2.2. and the other one that just cleans up - to be launched with `exec-shutdown`

Also, might be of interest: `HYPRLAND_NO_SD_VARS=1` [wiki](https://wiki.hyprland.org/Configuring/Environment-variables/#hyprland-environment-variables).

**Reply by sitedyno** [reply]
Thank you! I'll use the session service from examples for now, and start w/ `systemctl --user start --wait hyprland-session.service` from the tty. Working so far.
